crates/tree-sitter-essence/grammar.js

```diff
@@ -0,0 +1,598 @@
+module.exports = grammar ({
+  name: 'essence',
+
+  extras: $ => [
+    $.single_line_comment,
+    /\s/
+  ],
+
+  rules: {
+    program: $ => repeat(choice(
+      $.find_statement_list,
+      $.constraint_list,
+    )),
+
+    single_line_comment: $ => token(seq(
+      '$',
+      /.*/
+    )),
+
+    find_statement_list: $ => seq(
+      "find",
+      repeat($.find_statement)
+    ),
+
+    find_statement: $ => seq(
+      field("variable_list", $.variable_list),
+      ":",
+      field("domain", $.domain),
+      optional(",")
+    ),
+
+    variable_list: $ => seq(
+      $.variable,
+      optional(repeat(seq(
+        ",",
+        $.variable
+      )))
+    ),
+
+    constraint_list: $ => seq(
+      "such that",
+      repeat($.constraint)
+    ),
+
+    constraint: $ => seq(
+      optional($.not),
+      $.expression,
+      optional(",")
+    ),
+
+    not: $ => "!",
+
+    domain: $ => choice(
+      $.int_domain,
+      $.bool_domain
+    ),
+
+    int_domain: $ => prec.left(seq(
+      "int",
+      optional(seq(
+        "(",
+        $.range_list,
+        //eventually add in expressions here
+        ")"
+      ))
+    )),
+
+    bool_domain: $ => "bool",
+
+    range_list: $ => prec(2, seq(
+      choice(
+      $.lower_bound_range,
+      $.upper_bound_range,
+      $.closed_range,
+      $.integer
+      ),
+      optional(repeat(seq(
+        ",",
+        choice(
```

Comment:
instead of the four options of ranges, could you have done optional(), "..", optional() I wonder?